encoding.bs

```diff
@@ -94,37 +96,177 @@ bitwise AND by "&amp;", and bitwise OR by "|".
 
 <hr>
 
-<p>A <dfn id=concept-token>token</dfn> is a piece of data, such as a <a>byte</a> or
-<a>scalar value</a>.
+<p>An <dfn id=concept-stream export>I/O queue</dfn> is a type of <a>list</a>
+with <a>items</a> of a particular type (i.e. <a>bytes</a> or <a>scalar
+values</a>). <dfn id="end-of-stream" export>End-of-file</dfn> is a special
+<a>item</a> that can be present in <a for=/>I/O queues</a> of any type, and that
+signifies no more <a>items</a> are in the queue.
 
-<p>A <dfn id=concept-stream>stream</dfn> represents an ordered sequence of
-<a>tokens</a>. <dfn>End-of-stream</dfn> is a special
-<a>token</a> that signifies no more
-<a>tokens</a> are in the
-<a for=/>stream</a>.
+<p>To <dfn>create an I/O queue</dfn>, return a new <a for=/>I/O queue</a>
+containing an <a>end-of-file</a> item.
 
-<p>When a <a>token</a> is
-<dfn id=concept-stream-read for=stream>read</dfn> from a <a for=/>stream</a>,
-the first token in the stream must be returned and subsequently removed, and
-<a>end-of-stream</a> must be returned otherwise.
-<!-- this means read is blocking on e.g. networking activity;
-     SimonSapin thinks this is fine, blame him if not -->
+<div class=note>
+ <p>There are two ways to use an <a for=/>I/O queue</a>: in immediate mode, to
+ represent I/O data stored in memory, and in streaming mode, to represent data
+ coming in from the network. Immediate queues have <a>end-of-file</a> as their
+ last item, whereas streaming queues need not have it, and so their
+ <a for="I/O queue">read</a> operation might block.
+
+ <p>An immediate <a for=/>I/O queue</a> is to be created with the <a>create an
+ I/O queue</a> operation. A streaming <a for=/>I/O queue</a> is to be created
+ empty, and it is expected that new <a>items</a> will be
+ <a for="I/O queue">pushed</a> to it as data comes in from the network. When the
+ underlying network stream closes, an <a>end-of-file</a> item is to be
+ <a for="I/O queue">pushed</a> into the queue.
+
+ <p>Since reading from a streaming <a for=/>I/O queue</a> might block, streaming
+ <a for=/>I/O queues</a> are not to be used from an <a for=/>event loop</a>.
+ They are to be used <a>in parallel</a> instead.
+</div>
+
+<p>To <dfn id=concept-stream-read for="I/O queue" export>read</dfn> an
+<a>item</a> from an <a for=/>I/O queue</a> <var>ioQueue</var>, run these steps:
+
+<ol>
+ <li><p>If <var>ioQueue</var> is <a for=list>empty</a>, then wait until its
+ <a for=list>size</a> is at least 1.
+
+ <li><p>If the first item in <var>ioQueue</var> is <a>end-of-file</a>, then
+ return <a>end-of-file</a>.
+
+ <li><p>Otherwise, <a for=list>remove</a> the first item in <var>ioQueue</var>
+ and return it.
+</ol>
+
+<p>To read a number <var>number</var> of <a>items</a> from <var>ioQueue</var>,
```

Comment:
If we need this primitive it should have a dfn, no?

Reply:
Here I was trying to take algorithms defined to take "one or more tokens" and split them into two algorithms that took an item and a sequence. But maybe I should try and merge both overloads into one algorithm.